mdp/nodes/sfa_nodes.py

```diff
@@ -68,10 +72,37 @@ class SFANode(Node):
 
           You can even change this behaviour during training. Just set the
           corresponding switch in the `train` method.
+
+
+      ``handle_rank_deficit``
+          If ``True`` the `stop_train` method solves the SFA eigenvalue
+          problem in a way that is robust against linear redundancies in
+          the input data. This leads to rank deficit in the covariance
+          matrix, which usually yields a
+          SymeigException ('Covariance matrices may be singular').
+          There are several solving methods implemented, which you can
+          select by setting the _sfa_solver field, e.g.
+
+             sfa = SFANode()
+             sfa._sfa_solver = sfa._rank_deficit_solver_pca
+
+          Currently there are three solvers available:
+
+          _rank_deficit_solver_reg    - works by regularization
+          _rank_deficit_solver_pca - works by PCA
+          _rank_deficit_solver_svd    - works by svd
+
+            sfa = SFANode(handle_rank_deficit=True)
+
+          is equivalent to
+
+             sfa = SFANode()
+             sfa._sfa_solver = sfa._rank_deficit_solver_reg
+          
     """
 
     def __init__(self, input_dim=None, output_dim=None, dtype=None,
-                 include_last_sample=True):
+                 include_last_sample=True, handle_rank_deficit=False):
```

Comment:
As mentioned before, I'd prefer to name the solutions directly, i.e., an `eig_solve` (for example) parameter than takes the values of `'symeig'`, `'svd'`, etc. and allows to directly choose the preferred method. The default can either be `'symeig'` or whatever comes out ahead during benchmark testing.

Reply:
I suppose handle_rank_deficit should be turned off by default and should be trivial for a user to be turned on (using a default rank deficit solver), i.e. without looking into docs or code for available solver names.

I thought about having e.g. the reg-solver as default name, i.e. `rank_deficit_solver=__rank_deficit_solver_reg`, but then how would we have rank deficit solving turned off by default?

Reply:
Well if the user runs into the problem, then the error message will tell her how to deal with it either way: Switch `eig_solve` to a different value (and listing the available ones) or switching `handle_rank_deficit` to `True` - however, in case of the latter, which method will be used to handle the rank deficit and how can we choose from the multiple solutions on offer..? I still feel like the method should be explicitly named (and chosen) by setting this parameter.

Reply:
There is another issue to consider. The solvers need access to `self` in order to retrieve `_rank_threshold` and to write `rank_deficit`. So the solver methods are bound to the SFANode instance which would complicate to hand them as function pointers.
If we would call it like
`sfa = SFANode(output_dim=50, sfa_solver=SFANode._rank_deficit_solver_reg)`.
we would have to add code for manually binding the function to self (i.e. turning it into a method). I think this is possible in Python, but I would have to find out how it is done. I suspect this kind of thing might also require a Python2 vs Python3 distinction.
What about using keywords for solver names? I remember this pattern from several libraries, including SciPy and Tensorflow.
I.e. the doc would provide a list of solver names e.g. "reg", "svd", "pca", that are acceptable values. This way we could include values "none" and "default" or "auto". "auto" would insert the solver that performs best in our benchmarks (after we find a proper config to base benchmarks reliably...), while "none" would default to plain symeig.

```
    def __init__(self, input_dim=None, output_dim=None, dtype=None,
            include_last_sample=True, rank_deficit_handler="none"):
```
`sfa1 = SFANode(output_dim=50, rank_deficit_handler="svd")`
`sfa2 = SFANode(output_dim=50, rank_deficit_handler="auto")`



Using `sfa._sfa_solver = whatever` it would still be possible to insert a fully custom solver.

Reply:
> What about using keywords for solver names?

This is precisely what I was suggesting in my comments so that's obviously an excellent idea ;)

Except that it shouldn't be called `sfa_sover`, but something like `eig_solve` since it's solving the eigenvalue problem, not SFA itself.